### PR TITLE
Masks can fail IsMask, but still be applied as a mask

### PR DESCRIPTION
**Pull request by mcasperson**
A previous fix made the assumption that if a string is not a mask (i.e. `1.2.3-i` where `SemanticVersionMask.IsMask("1.2.3-i")` returns false), then it is not a mask. This is not correct: `1.2.3-i` is still applied as a mask in all functional cases in the existing implementation.

This PR pulls in the old implementation of `SemanticVersionMask` class and updates all the tests to verify the new behaviour against the old. It also addresses the cases where strings fail `IsMask`, but are still applied as masks.

See https://octopus.zendesk.com/agent/tickets/71996

### CODE REVIEW

File: source/Octopus.Versioning/Octopus/OctopusVersionMaskParser.cs [hunk 1]
```diff
@@ -49,7 +50,16 @@ public OctopusVersionMask Parse(string? version)
...
+
+            /*
+             * Watch out for this! A mask can return false for IsMask, but still succeed
+             * parsing here and proceed to apply the masking logic. See OctopusVersionMaskParserTests.IsMask
+             * for edge cases where masks fail IsMask but still work as masks.
+             *
+             * See SemanticVersionMask.ApplyMask in the test project for the original logic
+             * that has been replicated here for consistency.
+             */
+            if (!parsedMask.DidParse)
```
Comment (mcasperson) [inline source/Octopus.Versioning/Octopus/OctopusVersionMaskParser.cs:62]: This is the change that allows masks like `1.2.3-i` to be used, and retains compatibility with the old class.

File: source/Octopus.Versioning.Tests/Octopus/OctopusVersionMaskParserTests.cs [hunk 2]
```diff
@@ -63,48 +55,72 @@ public class OctopusVersionMaskParserTests
...
+            foreach (var prefix in VersionPrefixes)
+            {
+                var resultNewImplementationWithPrefix = OctopusVersionMaskParser.ApplyMask(prefix + mask, latestVersion != null ? new OctopusVersionParser().Parse(latestVersion) : null);
+                Assert.AreEqual(prefix + expected, resultNewImplementationWithPrefix.ToString());
+                Assert.AreEqual(prefix + expected, resultNewImplementationWithPrefix.OriginalString);
+            }
         }
 
+        /// <summary>
+        /// Assert that the new an old implementations fail with the same invalid input with whitespace around the prerelease field
```
Comment (MissedTheMark) [inline source/Octopus.Versioning.Tests/Octopus/OctopusVersionMaskParserTests.cs:105]: new **an** old ➡ new **and** old

**Review by MissedTheMark — APPROVED** [review]
Minor typo found in a test method summary, but nothing else.  I like the tests comparing the old and new functionality.

And you're right, that behaviour around .IsMask is definitely confusing.  I do appreciate you adding the comments explaining it.

File: source/Octopus.Versioning.Tests/Octopus/OctopusVersionMaskParserTests.cs (hunk 2 above)
Comment (mcasperson) [reply source/Octopus.Versioning.Tests/Octopus/OctopusVersionMaskParserTests.cs:105]: 👍 This has been cleaned up
